### PR TITLE
[IMPROVEMENT] Change bottom sheet lib

### LINKED ISSUE #759: Use reanimated-bottom-sheet
We switched action sheet lib months ago, but it's still not what we're looking for.
While native look and feel is great, the current lib isn't customizable.

https://github.com/osdnk/react-native-reanimated-bottom-sheet aims to implement [Material bottom sheet](https://material.io/design/components/sheets-bottom.html#standard-bottom-sheet) and it's using react-native-gesture-handler and react-native-reanimated to do it. Gestures and animations are great!

We should apply it on message actions and in messabox autocomplete container (for emojis, usernames, etc).

### PR DESCRIPTION
**Pull request by IlarionHalushka**
@RocketChat/ReactNative
Closes #759

This PR aims to replace Action Sheet with Reanimated Bottom Sheet on message actions.

As you could notice I am not a great designer so suggestions are welcome :)
big screen Iphone X: 
image

small screen Iphone SE: 
image

### CODE REVIEW

**Comment by diegolmello** [comment]
@IlarionHalushka Cool!!
We're not trying to copy Slack, but here's how they do it:


The idea is to show an amount of items and let the user swipe up if he wants to show the rest.
Also add the icons on the left.

For small screens: the tallest person in the world can have the smallest device, so he still needs a way to tap on the buttons :)
Leave the buttons in the same size and don't worry about it for now.

**Comment by diegolmello** [comment]
@IlarionHalushka this is how our design looks like for commands, autocomplete, etc:


Use your lib with our design and you're fine :)

**Comment by IlarionHalushka** [comment]
I updated the styles and added close on click-out. 

When Longpress performed on message:
image

When user scrolled the bottom action sheet from bottom to top:
image

**Comment by diegolmello** [comment]
@IlarionHalushka Looking so much better already!
Fix android build and I'll do a review.

**Comment by IlarionHalushka** [comment]
Improved the transitions: https://drive.google.com/file/d/1V4JmmbyzgIqmTaLjBfdjbLdzMbvPF8Q1/view?usp=sharing

File: app/containers/MessageActions.js [hunk 1]
```diff
@@ -264,79 +276,72 @@ export default class MessageActions extends React.Component {
...
+					</View>
+				</Touchable>
+			))}
+		</View>
+	);
+
 	render() {
 		return (
-			null
+			<BottomSheet
```
Comment (diegolmello) [inline app/containers/MessageActions.js]: A backdrop is missing.
You can use something like this: https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/develop/app/views/RoomsListView/SortDropdown.js#L121

File: app/views/RoomView/index.js [hunk 2]
```diff
@@ -131,6 +130,8 @@ export default class RoomView extends LoggedView {
 			} else {
 				EventEmitter.addEventListener('connected', this.handleConnected);
 			}
+
+			navigation.addListener('willBlur', () => this.hideActionSheet());
```
Comment (diegolmello) [inline app/views/RoomView/index.js]: I think this is not needed as RoomView will be removed from the stack, but test it.

If it's necessary, check if it's opened and move it to `componentWillUnmount`.
This is similar to `editCancel` and `replyCancel`.

File: ios/RocketChatRN.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -1684,20 +1693,32 @@
 				CLANG_CXX_LIBRARY = "libc++";
 				CLANG_ENABLE_MODULES = YES;
 				CLANG_ENABLE_OBJC_ARC = YES;
+				CLANG_WARN_BLOCK_CAPTURE_AUTORELEASING = YES;
```
Comment (diegolmello) [inline ios/RocketChatRN.xcodeproj/project.pbxproj]: I think you all these `CLANG*` are from your Xcode setup and shouldn't be commited.

File: app/views/RoomView/index.js [hunk 4]
```diff
@@ -490,15 +499,15 @@ export default class RoomView extends LoggedView {
...
 		} = this.props;
 		if (!navigation.isFocused()) {
 			return null;
 		}
 		return (
 			<React.Fragment>
-				{room._id && showActions
-					? <MessageActions room={room} user={user} />
+				{room._id
+					? <MessageActions room={room} user={user} setRef={ref => this.setRef(ref)} />
```
Comment (diegolmello) [inline app/views/RoomView/index.js]: Creating arrow functions on render is a bad practice and it should be avoided.
This creates functions on every render.

You should always be using the class function directly like this:
`setRef={this.setRef}`

File: ios/RocketChatRN.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -22,6 +22,7 @@
 		146834051AC3E58100842450 /* libReact.a in Frameworks */ = {isa = PBXBuildFile; fileRef = 146834041AC3E56700842450 /* libReact.a */; };
 		24A2AEF2383D44B586D31C01 /* libz.tbd in Frameworks */ = {isa = PBXBuildFile; fileRef = 06BB44DD4855498082A744AD /* libz.tbd */; };
 		2C800DF680F8451599E80AF1 /* libSafariViewManager.a in Frameworks */ = {isa = PBXBuildFile; fileRef = 1D3BB00B9ABF44EA9BD71318 /* libSafariViewManager.a */; };
+		3EAC15CC2271955100AD43B2 /* libRNReanimated.a in Frameworks */ = {isa = PBXBuildFile; fileRef = 3EAC15CB2271937800AD43B2 /* libRNReanimated.a */; };
```
Comment (diegolmello) [inline ios/RocketChatRN.xcodeproj/project.pbxproj]: Am I wrong or did you tried to add reanimated from imports and from pods?
If I'm right, it's just a matter of changing `Podfile` and lines like this one is trash.

File: android/app/build.gradle [hunk 6]
```diff
@@ -243,4 +244,4 @@ task copyDownloadableDepsToLibs(type: Copy) {
 }
 
 apply plugin: 'com.google.gms.google-services'
-com.google.gms.googleservices.GoogleServicesPlugin.config.disableVersionCheck = true
\ No newline at end of file
+com.google.gms.googleservices.GoogleServicesPlugin.config.disableVersionCheck = true
```
Comment (diegolmello) [inline android/app/build.gradle]: Unnecessary change

Comment (IlarionHalushka) [reply android/app/build.gradle]: this was made by Webstorm, returned back

File: ios/RocketChatRN.xcodeproj/project.pbxproj (hunk 5 above)
Comment (IlarionHalushka) [reply ios/RocketChatRN.xcodeproj/project.pbxproj]: I did link the library manually using Xcode, but not sure what all these lines are about... :(

File: app/views/RoomView/index.js (hunk 4 above)
Comment (IlarionHalushka) [reply app/views/RoomView/index.js]: did bind `this.setRef` and then used like you suggested `setRef={this.setRef}`

File: ios/RocketChatRN.xcodeproj/project.pbxproj (hunk 3 above)
Comment (IlarionHalushka) [reply ios/RocketChatRN.xcodeproj/project.pbxproj]: removed in latest commit

File: app/views/RoomView/index.js (hunk 2 above)
Comment (IlarionHalushka) [reply app/views/RoomView/index.js]: unfortunately the bottomsheet doesn't have function `onPositionChange` or something similar, but I did manage to do backdrop using `Animated.Code`

File: app/containers/MessageActions.js (hunk 1 above)
Comment (IlarionHalushka) [reply app/containers/MessageActions.js]: unfortunately the bottomsheet doesn't have function `onPositionChange` or something similar, but I did manage to do backdrop using `Animated.Code`

File: app/views/RoomView/index.js (hunk 2 above)
Comment (IlarionHalushka) [reply app/views/RoomView/index.js]: It's necessary, because when I open Bottom Action Sheet in RoomView and then go to Threads Single View (thread chat) then ActionSheet is still displayed there and same happens when changing RoomViews(chats)

File: app/containers/MessageActions.js [hunk 7]
```diff
@@ -264,79 +276,72 @@ export default class MessageActions extends React.Component {
...
-					this.handleDelete();
-					break;
-				default:
-					break;
-			}
-		}
-		const { actionsHide } = this.props;
-		actionsHide();
+	hideActionSheet() {
+		this.bottomSheetRef.current.snapTo(hiddenBottomSheet);
```
[the remark this answers is not in the document]
Comment (IlarionHalushka) [reply app/containers/MessageActions.js]: fixed

File: app/containers/MessageActions.js [hunk 8]
```diff
@@ -138,36 +161,25 @@ export default class MessageActions extends React.Component {
...
-	}
-
 	getPermalink = async(message) => {
 		try {
 			return await RocketChat.getPermalink(message);
 		} catch (error) {
 			return null;
 		}
-	}
+	};
```
[the remark this answers is not in the document]
Comment (IlarionHalushka) [reply app/containers/MessageActions.js]: removed

File: app/views/RoomView/index.js (hunk 2 above)
Comment (IlarionHalushka) [reply app/views/RoomView/index.js]: And also I tried to move it to `componentWillUnmount` but as we use StackNavigator - the view in not unmounted when opening Threads View or Settings view and bottom sheet stays constantly opened.

**Comment by pranavpandey1998official** [comment]
@IlarionHalushka  can you resolve conflicts

**Comment by pranavpandey1998official** [comment]
@IlarionHalushka the vibration are setting off immediately after opening the room and not after we long press ..

**Comment by pranavpandey1998official** [comment]
In my testing no button work

**Comment by pranavpandey1998official** [comment]
where did the separator go , It was nice as user was able to distinguish between boundaries of button 
WhatsApp Image 2019-05-11 at 6 27 00 PM

**Comment by pranavpandey1998official** [comment]
its showing start all the time (unstar is not showing up)

**Comment by pranavpandey1998official** [comment]
pin is not working

**Comment by IlarionHalushka** [comment]
**1.** `covers full screen screen in landscape` . fixed 
image

**2.** `where did the separator go , It was nice as user was able to distinguish between boundaries of button` - fixed.

**3.** `its showing start all the time (unstar is not showing up)` - currently I don't have any good idea how to change the titles because the bottomSheet is rendered once when the RoomView is opened. 

**4.** `pin is not working` - is not connected to BottomSheet, action is called but 400 Not Authorized response is returned from Back-end (same behaviour on current testflight app). [Issue](https://github.com/RocketChat/Rocket.Chat/issues/13454)
